meilisearch-http/src/backup.rs

```diff
@@ -0,0 +1,418 @@
+    let indexes = match crate::routes::index::list_indexes_sync(&data, &main_reader) {
+        Ok(indexes) => indexes,
+        Err(e) => {
+            fail_backup_process(backup_info, "listing indexes", e);
+            return ;
+        }
+    };
+
+    // create metadata
+    if let Err(e) = backup_metadata(&data, &tmp_dir_path, indexes.clone()) {
+        fail_backup_process(backup_info, "generating metadata", e);
+        return ;
+    }
+
+    // export settings, updates and documents for each indexes
+    for index in indexes {
+        let index_path = tmp_dir_path.join(&index.uid);
+
+        // create index sub-dircetory
+        if let Err(e) = create_dir_all(&index_path) {
+            fail_backup_process(backup_info, &format!("creating directory for index {}", &index.uid), e);
+            return ;
+        }
+
+        // export settings
+        if let Err(e) = backup_index_settings(&data, &main_reader, &index_path, &index.uid) {
+            fail_backup_process(backup_info, &format!("generating settings for index {}", &index.uid), e);
+            return ;
+        }
+
+        // export documents
+        if let Err(e) = backup_index_documents(&data, &main_reader, &index_path, &index.uid) {
+            fail_backup_process(backup_info, &format!("generating documents for index {}", &index.uid), e);
+            return ;
+        }
+
+        // export updates
+        if let Err(e) = backup_index_updates(&data, &update_reader, &index_path, &index.uid) {
+            fail_backup_process(backup_info, &format!("generating updates for index {}", &index.uid), e);
+            return ;
+        }
+    }
+
+    // compress backup in a file named `{backup_uid}.tar.gz` in `backup_folder`
+    if let Err(e) = crate::helpers::compression::to_tar_gz(&tmp_dir_path, &compressed_backup_folder(&backup_folder, &backup_info.uid)) {
+        fail_backup_process(backup_info, "compressing backup", e);
+        return ;
+    }
+
+    // update backup info to `done`
+    let resume = BackupInfo::new(
+        backup_info.uid,
+        BackupStatus::Done
+    );
+
+    resume.set_current();
+}
+
+pub fn init_backup_process(data: &web::Data<Data>, backup_folder: &Path) -> Result<BackupInfo, Error> {
+    create_dir_all(backup_folder).or(Err(Error::backup_failed()))?;
+
+    // check if a backup is already in progress
+    if let Some(resume) = BackupInfo::get_current() {
+        if resume.backup_already_in_progress() {
+            return Err(Error::backup_conflict())
+        }
+    }
+
+    // generate a new backup info
+    let resume = BackupInfo::new(
+        generate_uid(),
+        BackupStatus::Processing
+    );
+
+    resume.set_current();
+
+    {
+        let data = data.clone();
+        let backup_folder = backup_folder.to_path_buf();
+        let resume = resume.clone();
+        // run backup process in a new thread
+        thread::spawn(move || 
+            backup_process(data, backup_folder, resume)
+        );
+    }
```

Comment:
Why is this in it's own scope? I don't see any lock or anything like that.

Reply:
@Kerollmops because of `let resume = resume.clone();` which is a `clone`, I can remove the scope and rename it `resume_cloned`

Reply:
Ho! I see, yeah it would be indeed better I think 👍

Reply:
changed it